Implemented adaptive squashing

dholzmueller opened the pull request:
WIP

dholzmueller commented:
Here is an implementation of the robust scaling + smooth clipping from my paper (https://arxiv.org/abs/2407.04491). It could be used as a robust numerical preprocessing methods for neural network based methods.

Some considerations:
- which name do we want to use for the class? Currently it is AdaptiveSquashingTransformer.
- I did not check the rendering of the documentation. Do we need to link the class from somewhere in the documentation?
- If the column passed in transform() has a different name than in fit_transform(), the output column still gets the name of the column passed in fit_transform(). I copied this behavior from StringEncoder, is that good? (See also the todo in the corresponding tests.)

dholzmueller commented on skrub/_adaptive_squashing.py:
```diff
@@ -0,0 +1,196 @@
+        """
+
+        values_np = sbd.to_numpy(X).astype(np.float32)
+        result = np.copy(values_np)
```
Are you sure that it is a copy of X? to_numpy() in pandas does not guarantee to make a copy unless you set copy=True, which is not set as far as I can see. Also `del X` is not possible because `X` is used afterwards.

jeromedockes replied:
I think that in `np.ndarray.astype` the default for `copy` is `True`

`del X` probably wouldn't change memory usage because the caller likely holds a reference to X as well. I'm guessing the `del y` used elsewhere in this PR is not to decrease the reference count but to make it clear that y is intentionally not used

dholzmueller commented:
> Also, I think having a smallish example that shows when and where to use your class would be great. WDYT?

Sounds good, where would you put it?

dholzmueller commented on skrub/_adaptive_squashing.py:
```diff
@@ -0,0 +1,196 @@
+    ----------
+    For a detailed description of the method, see
+    `Better by default: Strong pre-tuned MLPs and boosted trees on tabular data
+    <https://arxiv.org/abs/2407.04491>`_ by Holzmüller, Grinsztajn, Steinwart (2024).
```
Yep, I added one.

dholzmueller commented on skrub/_adaptive_squashing.py:
```diff
@@ -0,0 +1,196 @@
+            raise RejectColumn(f"Column {self.col_name_} is not numeric.")
+
+        if not self.col_name_:
+            self.col_name_ = "adaptive_squashed"
```
done

dholzmueller commented on skrub/_adaptive_squashing.py:
```diff
@@ -0,0 +1,196 @@
+    NaN values will be preserved.
+
+    The output feature will be named ``{col_name}`` if the series has a name,
+    and ``adaptive_squashed`` if it does not.
```
ok, removed it

dholzmueller commented on skrub/_adaptive_squashing.py:
```diff
@@ -0,0 +1,196 @@
+    proposed for tabular neural networks by
+    David Holzmüller, Léo Grinsztajn, and Ingo Steinwart,
+    Better by default: Strong pre-tuned MLPs and boosted trees on tabular data,
+    Advances in Neural Information Processing Systems, 2024.
```
done

dholzmueller commented:
I addressed everything except from the example now.

dholzmueller commented:
I removed the test that tested if a column with name `None` after transformation still had name `None`, since it failed for polars. I assume this is not relevant?

jeromedockes commented on CHANGES.rst:
```diff
@@ -27,6 +27,12 @@ Lemaitre <glemaitre>`, :user:`Adrin Jalali <adrinjalali>`, :user:`Olivier Grisel
 <ogrisel>` and others participated through many discussions in defining the
 requirements and the public API.
 
+- Added the `AdaptiveSquashingTransformer` that
```
I think there is a formatting issue:

![screenshot_2025-04-15T13:32:29+02:00](https://github.com/user-attachments/assets/91926215-8ca8-4047-8c8f-90fcf7581a97)


the lines below the `-` need to be indented to still be part of that bullet point, and the last line is too indented which turns it into a block quote. yes, restructured text is an acquired taste. a couple of things that can help:

- https://docutils.sourceforge.io/docs/user/rst/quickref.html
- in the pr page in the checks at the bottom click on "check the rendered docs here!"
- or locally go in the `/doc/` folder and there `make html-noplot` to render the doc without actually running the examples. you may need to first `pip install ".[doc]"`

jeromedockes replied:
also you can do 
```
:class:`AdaptiveSquashingTransformer`
```

that will turn it into a link to the documentation

jeromedockes commented on doc/reference/index.rst:
```diff
@@ -210,6 +210,7 @@ Cleaning a dataframe
    :nosignatures:
 
    deduplicate
+   AdaptiveSquashingTransformer
```
how about `HolzmüllerTransformer` or `AdaptiveScaler`

dholzmueller replied:
Well, I don't think users would like to write an ü. AdaptiveScaler is missing the squashing/clipping aspect. What about SquashScaler/ScaleSquasher?

jeromedockes replied:
that's true. ok we can go for Squash(ing)Scaler for now and we still have until the next release to change our mind

Vincent-Maladiere replied:
what does @GaelVaroquaux think?

jeromedockes commented on skrub/_adaptive_squashing.py:
```diff
@@ -0,0 +1,207 @@
+    ----------
+    max_absolute_value : float, default=3.0
+        Maximum absolute value that the transformed data can take.
+    lower_quantile_alpha : float, default=0.25
```
could it be `lower_quantile`?

dholzmueller replied:
This was my first draft, but I wanted to distinguish between the quantile (value in \bbR) and the probability level (in [0, 1]) corresponding to the quantile (could be called quantile_level or quantile_alpha).

jeromedockes replied:
that makes sense. OTOH it wouldn't make sense to pass the actual quantile value as an argument so people can probably guess what it means. or this parameter is called `q` in numpy and pandas, maybe this could be an option too

dholzmueller replied:
maybe q_lower and q_upper?

jeromedockes commented on skrub/_adaptive_squashing.py:
```diff
@@ -0,0 +1,207 @@
+        del y
+
+        if not isinstance(self.max_absolute_value, numbers.Number):
+            raise ValueError(
```
I guess a TypeError would be more appropriate if it is the wrong type

jeromedockes commented on skrub/_adaptive_squashing.py:
```diff
@@ -0,0 +1,207 @@
+        if not sbd.is_numeric(X):
+            raise RejectColumn(f"Column {self.col_name_} is not numeric.")
+
+        eps = 1e-30  # todo: expose?
```
```suggestion
        eps = 1e-30
```

no one will set it so we shouldn't expose it but just make sure we picked a good value. if needed you can use stuff like `np.finfo('float32').tiny`

jeromedockes reviewed (commented):
a first few comments, I'll add more. thanks this is great!! :rocket:

jeromedockes commented on skrub/_adaptive_squashing.py:
```diff
@@ -0,0 +1,207 @@
+        if len(finite_values) > 0:
+            self.median_ = np.median(finite_values)
+            quantiles = [
+                np.quantile(finite_values, q)
```
not important but wouldn't `np.quantile(finite_values, [...])` work? (you can ask for multiple quantiles in one call)

jeromedockes commented on skrub/_adaptive_squashing.py:
```diff
@@ -0,0 +1,207 @@
+            self.median_ = 0.0
+            self.scale_ = 1.0
+
+        self._is_fitted = True
+        self.n_components_ = 1
```
```suggestion
```

I don't think we need those

jeromedockes commented on skrub/_adaptive_squashing.py:
```diff
@@ -0,0 +1,207 @@
+        # not the name of X
+        return self._post_process(X, values_np, self.col_name_)
+
+    def _post_process(self, X, result, name):
```
if this function is only called in 1 place we can just inline those 2 lines and not define a function for it

Vincent-Maladiere commented:
> Sounds good, where would you put it?

@dholzmueller, good question, I'm not sure. What would be the best way to illustrate its usefulness? Do you have a small benchmark or experiment that could fit into an example? Otherwise we might need to be creative

dholzmueller commented:
We could try it on a dataset with an MLP and see if it improves something. Don't know what would be the best one to do this.

dholzmueller commented:
Updates:
- renamed the class to SquashingScaler
- made it work for dataframes and not only series by using OnEachColumn with the old class
- renamed the args to lower_quantile and upper_quantile
- incorporated minor suggestions
- added an example using the SquashingScaler

jeromedockes commented on skrub/_squashing_scaler.py:
```diff
@@ -0,0 +1,341 @@
+        return result
+
+
+class SquashingScaler(BaseEstimator, TransformerMixin):
```
even though it wouldn't make a difference here normally we would put the transformermixin first because we want methods to be looked up there before looking in baseestimator

```suggestion
class SquashingScaler(TransformerMixin, BaseEstimator):
```

jeromedockes commented on skrub/_squashing_scaler.py:
```diff
@@ -0,0 +1,341 @@
+        self: SquashingScaler
+            The current object.
+        """
+        self.tfm_ = SingleColumnSquashingScaler(
```
maybe we could make the transformer private by renaming it to `self._transformer` and exposing its learned attributes directly by copying them to the outer object's dict with something like

```python
if isinstance(self._transformer, OnEachColumn):
    self.median_ = [t.median_ for t in self._transformer.transformers_.values()]
    self.scale_ = [t.median_ for t in self._transformer.transformers_.values()]
else:
    self.median_ = self._transformer.median_
    self.scale_ = self._transformer.scale_
```

jeromedockes replied:
(in this case we could add a section like

```
    Attributes
    ----------
    median_ : float or list of float
        median of the training data

    scale_ : ...
```

just below the 'parameters' section in the docstring )

jeromedockes reviewed (commented):
Great, thanks a lot @dholzmueller . I just have a small comment on the fitted attributes but LGTM otherwise :)

dholzmueller commented:
@GaelVaroquaux was suggesting to implement SquashingScaler directly instead of with the indirection through SingleColumnSquashingScaler. I'll need to check how easy that is...

GaelVaroquaux commented:
> that's an option too, I'll let you decide!

Users (and non advanced programmers) struggle with indirections.

Your point Jerome is about memory overhead? Do you think that 1) they will be really a problem 2) they end up being avoided, because the output of this transformer is likely going to be fed to a scikit-learn learner?

jeromedockes commented:
> Your point Jerome is about memory overhead?

yes because we go from columnar format, to one contiguous array, then back to columnar (dataframe format). that is the case even if the output dataframe is then passed to an estimator.

> 1) they will be really a problem

not necessarily but just in case it is easy, it is always nice to avoid an extra copy, especially when it requires a much larger contiguous buffer than any column in the input dataframe does. 

the main reason I mention it is because the fitting of each column as currently implemented uses python control flow anyway (to handle corner cases such as a column is all empty, the quantiles are equal, or the column is constant), so I suspect it will not be easy to express in a "vectorized" way on one numpy array. so if we end up with a loop over columns anyway, let's avoid building the full array. 

if it makes things more complicated you can ignore my comment

GaelVaroquaux commented on skrub/_squashing_scaler.py:
```diff
@@ -0,0 +1,341 @@
+from ._on_each_column import OnEachColumn, RejectColumn, SingleColumnTransformer
+
+
+class SingleColumnSquashingScaler(SingleColumnTransformer):
```
I seems to me that the logic could be implemented well at the level of multiple columns. Nan handling can be done using np.nanmean and np.nanquantile.

I think that this would be beneficial

Vincent-Maladiere commented:
Update: this is now ready for review. This transformer behaves like any scikit-learn preprocessor, taking a pandas or polars dataframe or numpy array as input, and yielding a numpy array. This can be used within the TableVectorizer or ApplyOnCols without additional logic.

Vincent-Maladiere commented:
Yes, totally. I was trying to minimize the differences between sklearn and skrub scalers APIs, but we definitely can.

rcap107 commented on skrub/_squashing_scaler.py:
```diff
@@ -0,0 +1,304 @@
+
+def _infinite_sign(X):
+    """Replace infinite values with NaN and return their sign."""
+    if (inf_sign := np.isinf(X)).any():
```
I find the name `inf_sign` confusing, I read it as "this is the sign used as infinte", not as the location of the inf 

I'd rather have `inf_mask` instead, I think it conveys better what this is supposed to do.

rcap107 commented on skrub/_squashing_scaler.py:
```diff
@@ -0,0 +1,304 @@
+    return X
+
+
+def _soft_clip(X, max_absolute_value, inf_sign):
```
here the docstring shows `mask_inf` instead of `inf_sign`, and I prefer `mask_inf` out of the two

rcap107 commented on skrub/_squashing_scaler.py:
```diff
@@ -0,0 +1,304 @@
+
+
+class SquashingScaler(OneToOneFeatureMixin, TransformerMixin, BaseEstimator):
+    r"""Perform robust centering and scaling of followed by soft clipping.
```
Maybe the sentence is missing something? 

```suggestion
    r"""Perform robust centering and scaling followed by soft clipping.
```

rcap107 commented on skrub/_squashing_scaler.py:
```diff
@@ -0,0 +1,304 @@
+    >>> import pandas as pd
+    >>> import numpy as np
+    >>> from skrub import SquashingScaler
+    >>> X = pd.DataFrame(dict(a=[-100.0, 0.0, 8.0, 9.0, 10.0, 20.0, 100.0, np.nan]))
```
maybe add inf to the example to show what happense

rcap107 replied:
there should also be an example showing that when the input column is 0, the transform is also converting it to 0 regardless of the values (if I'm understanding what the code is doing)

Vincent-Maladiere replied:
It's more general than that: if the column is constant, then the scaler fills it with zeros. I can add an example for all 3 cases.

rcap107 commented on skrub/tests/test_squashing_scaler.py:
```diff
@@ -0,0 +1,145 @@
+        tfm.fit_transform(X)
+
+
+def test_squashing_scaler_known_values(df_module):
```
please add a comment explaining what this is for

rcap107 reviewed (commented):
Overall, I think that the code in this PR is not documented well enough, as there are a lot of parts that "do stuff" in a non-transparent way, and that would benefit from being commented more. 

Besides being a problem to maintain (I had to try and reverse engineer most of the code and the tests to figure out how it worked), I think that the scaling could have some unexpected results that may surprise users, and that are not documented or explained properly in the docstrings or in the code itself.

Vincent-Maladiere commented on skrub/_squashing_scaler.py:
```diff
@@ -0,0 +1,304 @@
+            ensure_all_finite=False,
+        )
+        X, inf_sign = _infinite_sign(X)
+
```
Ok, will do

rcap107 commented:
> Hey @rcap107 thanks for reviewing. I iterated on the documentation; let me know what you think

It's much better, thanks a lot! I fixed a small typo, but aside from that I think we can merge this

dholzmueller commented:
Interface looks good to me! Did you test what happens if you pass in an all-nan column?

GaelVaroquaux commented:
I'm waiting for @Vincent-Maladiere 's reply to @dholzmueller 's above, and then merge

Vincent-Maladiere commented:
> Interface looks good to me! Did you test what happens if you pass in an all-nan column?

Good question. We get a column full of nans, since they are left untouched. Is that what you would expect?

Reproducer:
```python
import numpy as np
import skrub
import pandas as pd

skrub.SquashingScaler().fit_transform(
    pd.DataFrame(dict(a=[np.nan, np.nan]))
)
```

dholzmueller commented:
Yes, that is what I would expect. Just wasn't sure if it's going to work with the nanmax/nanmin operations.

dholzmueller commented:
Just a thought, having quantile_range as a tuple instead of two separate values might make it harder for people to tune it in some hyperparameter optimization. I think it's probably not an important hyperparameter to tune, though.

rcap107 commented:
> Just a thought, having quantile_range as a tuple instead of two separate values might make it harder for people to tune it in some hyperparameter optimization. I think it's probably not an important hyperparameter to tune, though.

This can be addressed in another PR

dholzmueller commented:
> This can be addressed in another PR
you mean, another one for the same release? because it's not as convenient to change the parametes after release, right?

dholzmueller commented:
okay, then let's leave it like it is.

dholzmueller commented:
I think the NaN thing is not an issue, since Vincent tested it and it worked.

GaelVaroquaux commented:
Merged!

Thanks everyone
